pilot/pkg/controller/workloadentry/workloadentry_controller.go

```diff
@@ -243,54 +244,76 @@ func (c *Controller) processNextWorkItem() bool {
 		return true
 	}
 
-	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime)
+	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime, workItem.autoCreated)
 	c.handleErr(err, item)
 	return true
 }
 
 func setConnectMeta(c *config.Config, controller string, conTime time.Time) {
+	if c.Annotations == nil {
+		c.Annotations = make(map[string]string)
+	}
 	c.Annotations[WorkloadControllerAnnotation] = controller
 	c.Annotations[ConnectedAtAnnotation] = conTime.Format(timeFormat)
 	delete(c.Annotations, DisconnectedAtAnnotation)
 }
 
 func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) error {
-	if !features.WorkloadEntryAutoRegistration || c == nil {
+	if c == nil {
 		return nil
 	}
-	// check if the WE already exists, update the status
-	entryName := autoregisteredWorkloadEntryName(proxy)
+	var entryName string
+	var autoCreate bool
+	if features.WorkloadEntryAutoRegistration &&
+		proxy.Metadata.AutoRegisterGroup != "" {
+		entryName = autoregisteredWorkloadEntryName(proxy)
+		autoCreate = true
+	} else if features.WorkloadEntryHealthChecks &&
+		proxy.Metadata.ProxyConfig != nil && proxy.Metadata.ProxyConfig.ReadinessProbe != nil {
```

Comment:
Looking for `ProxyConfig.ReadinessProbe` in the metadata seems concerning to me. The control plane shouldn't care what the probe is - just if it should wait until its been marked "Ready" before using the WE or not (which the annotation does)

Reply:
Sorry, can you explain what would be the alternative to checking `proxy.Metadata.ProxyConfig != nil && proxy.Metadata.ProxyConfig.ReadinessProbe != nil` ?

I do this check to filter out early connected proxies that correspond to `Pods`.

Should I drop this check altogether and simply try to find `WorkloadEntry` ?

Reply:
What confused me is why `WorkloadEntry` has a `proxy.istio.io/config` annotation with the readiness probe in it... How does that actually end up on the VM or get used? My understanding is the VM knows nothing about the WorkloadEntry?

How I would expect the code:
```
if isVM() {
  if _,f := annotations[WorkloadEntryHealthCheckAnnotation]; !f { return // no health checks, no need to do anything }
  entryName = ...
}
```

For IsVM, I think the change you mentioned to explicitly declare it in NodeMetadata below is likely ideal?

Reply:
@howardjohn Hey

> What confused me is why `WorkloadEntry` has a `proxy.istio.io/config` annotation with the readiness probe in it...

You're right that Istio doesn't provide a command anymore to generate configuration for the Istio Sidecar based on a `WorkloadEntry` rather than on a `WorkloadGroup`.

Yeah, I can remove `proxy.istio.io/config` from the example to avoid confusion.

The workflow that is possible right now:
* a user creates a `WorkloadGroup`
* then uses `istioctl x workload entry` command to generate configuration for the sidecar
* then manually creates  a `WorkloadEntry` to represent a single VM

> My understanding is the VM knows nothing about the `WorkloadEntry`?

Correct.

At the moment, the only link possible is if a user will set environment variable `POD_NAME` to the name of `WorkloadEntry` he/she created manually.

>   if _,f := annotations[WorkloadEntryHealthCheckAnnotation]; !f { return // no health checks, no need to do anything }

Ok, I will rework to check for `annotations[WorkloadEntryHealthCheckAnnotation]`. 

> For IsVM, I think the change you mentioned to explicitly declare it in NodeMetadata below is likely ideal?

Since I've been focused on the VM use case, I proposed to name this field `WORKLOAD_ENTRY` and use it only for `WorkloadEntry`(s).

But maybe you prefer a neutral field that could be used for either `Pod` name or `WorkloadEntry` name?

On the downside, it won't be enough to distinguish a VM.

Reply:
Use WORKLOAD_ENTRY, SGTM

Reply:
added `WORKLOAD_ENTRY` and  started checking `WorkloadEntryHealthCheckAnnotation`.

@howardjohn PTAL